src/shims/sync.rs

```diff
@@ -454,135 +456,325 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                // is any.
+                if let Some(writer) = this.rwlock_dequeue_writer(id) {
+                    this.unblock_thread(writer)?;
+                    this.rwlock_writer_set(id, writer);
                 }
             }
             Ok(0)
-        } else if writers != 0 {
-            let reader_blockset = rwlock_get_or_create_reader_blockset(this, rwlock_op)?;
+        } else if Some(active_thread) == this.rwlock_writer_remove(id) {
+            // The thread was a writer.
+            //
             // We are prioritizing writers here against the readers. As a
             // result, not only readers can starve writers, but also writers can
             // starve readers.
-            if let Some(_writer) = this.unblock_some_thread(writer_blockset)? {
-                assert_eq!(writers, 1);
+            if let Some(writer) = this.rwlock_dequeue_writer(id) {
+                // Give the lock to another writer.
+                this.unblock_thread(writer)?;
+                this.rwlock_writer_set(id, writer);
             } else {
-                rwlock_set_writers(this, rwlock_op, Scalar::from_u32(0))?;
-                let mut readers = 0;
-                while let Some(_reader) = this.unblock_some_thread(reader_blockset)? {
-                    readers += 1;
+                // Give the lock to all readers.
+                while let Some(reader) = this.rwlock_dequeue_reader(id) {
+                    this.unblock_thread(reader)?;
+                    this.rwlock_reader_add(id, reader);
                 }
-                rwlock_set_readers(this, rwlock_op, Scalar::from_u32(readers))?
             }
             Ok(0)
         } else {
-            throw_ub_format!("unlocked an rwlock that was not locked");
+            throw_ub_format!("unlocked an rwlock that was not locked by the active thread");
         }
     }
 
     fn pthread_rwlock_destroy(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        if rwlock_get_readers(this, rwlock_op)?.to_u32()? != 0
-            || rwlock_get_writers(this, rwlock_op)?.to_u32()? != 0
-        {
+        let id = rwlock_get_or_create_id(this, rwlock_op)?;
+
+        if this.rwlock_is_locked(id) {
             throw_ub_format!("destroyed a locked rwlock");
         }
 
-        rwlock_set_readers(this, rwlock_op, ScalarMaybeUndef::Undef)?;
-        rwlock_set_writers(this, rwlock_op, ScalarMaybeUndef::Undef)?;
-        rwlock_set_reader_blockset(this, rwlock_op, ScalarMaybeUndef::Undef)?;
-        rwlock_set_writer_blockset(this, rwlock_op, ScalarMaybeUndef::Undef)?;
+        rwlock_set_id(this, rwlock_op, ScalarMaybeUndef::Undef)?;
+
+        Ok(0)
+    }
+
+    fn pthread_condattr_init(&mut self, attr_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let default_clock_id = this.eval_libc("CLOCK_REALTIME")?;
```

Comment:
Why is this the default clock?

Reply:
I found this in the [documentation](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_condattr_setclock.html):

> The default value of the clock attribute shall refer to the system clock.

Have I mixed up the locks and `CLOCK_REALTIME` is not the system clock?

Reply:
`CLOCK_REALTIME` is used to implement `SystemClock`, so this makes sense. Please add a comment saying that this is specified to be the default.